llext: add support for STT_OBJECT relocations

lyakh opened the pull request:
These relocations are needed e.g. when an LLEXT is built from multiple files and they share data objects.

lyakh committed:
llext: add support for STT_OBJECT relocations

These relocations are needed e.g. when an LLEXT is built from
multiple files and they share data objects.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

teburd commented:
Would it be difficult to add a testable extension that covers we support global object linking? I think its worth adding if you don't mind spending the few extra minutes making it.

teburd commented:
Thank you for adding a test #72383

lyakh commented:
> Would it be difficult to add a testable extension that covers we support global object linking? I think its worth adding if you don't mind spending the few extra minutes making it.

@teburd #72383 and the result is interesting. Firstly - yes, I made it a separate PR on purpose to not slow down this one, while we polish the test ;-) Secondly - turns out that *currently* even the simple "object_ext" test is silently "green-failing" (TM) at least on Xtensa. It outputs:
> DEBUG   - QEMU (291790): I: Loaded extension object
> DEBUG   - QEMU (291790): number: 0
> DEBUG   - QEMU (291790): number, updated: 0
> DEBUG   - QEMU (291790): string: (null)

My new multi-file test from that PR is producing a similar 0-result. With my this PR I'm getting:
> DEBUG   - QEMU (288615): I: Loaded extension object
> DEBUG   - QEMU (288615): number: 42
> DEBUG   - QEMU (288615): number, updated: 0
> DEBUG   - QEMU (288615): string: hello

and
> DEBUG   - QEMU (288593): I: Loaded extension multi_file
> DEBUG   - QEMU (288593): initial: local 66 plus external 24 equals 90
> DEBUG   - QEMU (288593): updated: local 24 plus external 66 equals 90

respectively. That result is for llext.simple.writable and for modules_enabled_writable. However, for modules_enabled_writable_relocatable it's worse:
> DEBUG   - QEMU (288606): I: Loaded extension multi_file
> DEBUG   - QEMU (288606): initial: local 0 plus external 24 equals 24
> DEBUG   - QEMU (288606): updated: local 24 plus external 0 equals 24

and
> DEBUG   - QEMU (288606): I: Loaded extension object
> DEBUG   - QEMU (288606): number: 24
> DEBUG   - QEMU (288606): number, updated: 0
> DEBUG   - QEMU (288606): string: (null)

So that case still has to be fixed, and presumably twister should be updated to recognise failures.

teburd commented:
It's completely fine that its separate, agreed no need to slow down, but definitely a good practice to prevent regression and ensure that yes, we really did add the feature and we can prove it.

It's interesting that it doesn't fail today, so maybe some asserts in the extension are needed to ensure failure is caught, or some added call to tell the calling ztest (beyond an assert) that it failed (return an errno perhaps?)

pillo79 commented:
> It's interesting that it doesn't fail today, so maybe some asserts in the extension are needed to ensure failure is caught, or some added call to tell ztest its failed, or perhaps changing the signature of the extension test_entry to return integer codes where -errno is signaling error.

The ztest control macros are already available; however, enforcing full compliance exposes a few issues such as these that need to be tackled by fixing the linking code. That's my end goal... as soon as I can get to it :slightly_smiling_face:

lyakh commented:
> It's interesting that it doesn't fail today, so maybe some asserts in the extension are needed to ensure failure is caught, or some added call to tell ztest its failed, or perhaps changing the signature of the extension test_entry to return integer codes where -errno is signaling error.

@teburd @pillo79 How about this: after this one and #72383 are merged, we (me or somebody else) make a PR adding those checks. That PR will fail twister tests, so it won't be merged. So we'll keep fixing linking until it can be merged? Or we can add several such PRs exposing individual errors.

pillo79 commented:
> > It's interesting that it doesn't fail today, so maybe some asserts in the extension are needed to ensure failure is caught, or some added call to tell ztest its failed, or perhaps changing the signature of the extension test_entry to return integer codes where -errno is signaling error.
> 
> @teburd @pillo79 How about this: after this one and #72383 are merged, we (me or somebody else) make a PR adding those checks. That PR will fail twister tests, so it won't be merged. So we'll keep fixing linking until it can be merged? Or we can add several such PRs exposing individual errors.

That is an interesting approach! :+1:
